### ISSUE TITLE
[Proposal] Add an `LARGE_ADDRESS_AWARE` switch to the .NET compiler

### ISSUE THREAD

**Issue by Unknown6656** [dotnet/roslyn#9209]
I sometimes write applications and/or libraries which need the `LARGE_ADDRESS_AWARE`-flag set. I usually solve this by calling `editbin.exe` as postbuild-event.

Would be an good idea to add a corresponding switch directly to the C#/F#/.... compiler?
It would have the following pros:
- one could save a few miliseconds/seconds of compilation time and postbuild-event calls
- it is easier to use for regular users (when properly integrated into VisualStudio)
- would not influence the compiler speed a lot - as it is only one argument which must be parsed and one bit which must be written

_Or am I the sole one appending the `editbin`-postbuild to every second project of mine?_

**Comment by sharwell** [comment]
This is already implemented:

https://github.com/dotnet/roslyn/blob/0846ba67968d5bec15f5119b99849497ff987f49/src/Compilers/Core/Portable/Compilation/Compilation.cs#L1644-L1646

* Use **X86** platform to get a binary that runs as a 32-bit image and is not large address aware
* Use **Any CPU**+**Prefer32Bit** to get a binary that runs as a 32-bit image and is large address aware

**Comment by AArnott** [comment]
Thanks, @sharwell. That's a very simple workaround. 

How does it work when you're compiling for .NET Native though, where one *must* choose a CPU architecture rather than "any cpu"?

**Comment by tannergooding** [comment]
I agree, there should be a way to explicitly set these PE header flags as part of the compilation.

**Comment by sharwell** [comment]
@AArnott never used .NET Native. Sounds like a bug should be filed there to work with Any CPU outputs.

**Comment by tannergooding** [comment]
@sharwell, it also seems like a bug that you cannot set LAA for explicit x86, but it magically gets set for AnyCPU/Prefer32Bit.

There are native libraries you could potentially be interoping with that are not LAA, and will fail with that assumption.

There are also cases where you are explicitly targeting x86 because AnyCPU is invalid (explicit x86 prevents the binary from being loaded in a 64-bit processes, AnyCPU+Prefer32Bit does not) but where you may still want LAA.

**Comment by AArnott** [comment]
I'm not sure if .NET Native has a bug or not. I was just wondering about that scenario.

+1 to @tannergooding's points. We compile with x86 sometimes because we know we'll be loading x86 native binaries, so x64 is *not* an option. The compiler/build knows this and will actually emit warnings if we reference x86 assemblies while targeting AnyCPU ourselves. So yes, there should be a way to require x86 and still be LAA.

**Comment by monty241** [comment]
From the header I understand that this also might affect dll libraries. We create various vsto applications which should ship as anycpu if possible since some users have excel, word or outlook 32-bit and some 64-bit. 

Is there a good way to have these libraries also run in Excel versions with LAA and use the memory advantage, without loosing the option to use them with 64-bit office?

**Comment by AraHaan** [comment]
I agree, I also load up an dll (using reflection) made in C++ /clr and using AnyCPU+Prefer32Bit on an 64 bit OS with all the other .NET Assemblies AnyCPU and that is then loaded as an 64 bit process, now when that c++/clr dll is loaded using reflection (as a 32 bit dll) it then fails.

So then I changed the exe to be ``X86`` however because like was said above I think the compiler should set ``/largeaddressaware`` as a switch that we pass in instead.

**Comment by AraHaan** [comment]
Because of the fact that I made dotnet/project-system#6829 I would like to take a crack at this for the next compiler version, then an option can be added to the dotnet/Sdk respositroy, so then dotnet/project-system#6829 can then be done.

**Comment by AraHaan** [comment]
I see the latest C# compiler has the needed stuff in https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Compilation/Compilation.cs#L2041-L2043

However where is the Visual Basic Compiler's copy for it, and where do I go to add compilation options to the compiler?

**Comment by Youssef1313** [comment]
> I see the latest C# compiler has the needed stuff in https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Compilation/Compilation.cs#L2041-L2043
> 
> However where is the Visual Basic Compiler's copy for it, and where do I go to add compilation options to the compiler?

The Compilation.cs file you're referencing is both for C# and VB compilers. This is the `abstract` base class that's used by both languages. See:

https://github.com/dotnet/roslyn/blob/f735d0a84a9b3b8c9e524285a7fc6b4648669471/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L41

https://github.com/dotnet/roslyn/blob/f735d0a84a9b3b8c9e524285a7fc6b4648669471/src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb#L36-L37

**Comment by AraHaan** [comment]
Sweet so now all I need to do is figure out on adding /laa switch to both compilers and then having the thing call both methods in that Compilation class when it is specified explicitly.

Now for users targeting AnyCPU+Prefer32Bit should it call into that new method then to keep ABI as much as possible or should LAA only be set when adding the switch to the compiler instead?

**Comment by AraHaan** [comment]
After thinking more on this, I think LAA should be set unconditionally for every target (since it does it anyway for x64 too so I see why not set it in every target) I can't think of places where setting it at all would break things.

But before I do @tannergooding can you think of any places where changing the compiler to set LAA conditionally could break any targets at all?

**Comment by michaelcfanning** [comment]
If we introduce /largeaddressaware, the problem is that we have differing current default behaviors, /largeaddressaware- for requires 32-bit, /largeaddressaware+ otherwise. We could preserve this behavior in your change, however, and still allow users to explicitly provide /largeaddressaware+ for the latter scenario. I think that's probably the right choice.

Because otherwise, as you clearly understand, a compiler update might unexpectedly opt 32-bit code into LAA causing runtime issues that aren't caught in testing. And I'm not sure an environment variable can provide a good mitigation. For one thing, I think the OS loader would actually need to honor that variable. We'd also need to think about whether this environment variable was inherited by child processes. That could be bad in many cases, for correctness but also because it could disable security features such as high entropy virtual addresses for ASLR.

Speaking of which, the csc.exe compiler obviously already supports /highentropyva and this security setting is not meaningful unless the large address aware bit is also set. And so any command-line with /highentropyva+ and /largeaddressaware- should throw an error (because you won't in fact enable the high entropy security feature in this case). This problem exists today, i.e., people can compile managed binaries with high entropy enabled but missing the LAA bit. I probably should file an issue on that and reference this one.
